File: 백준/Silver I/2583. 영역 구하기/영역 구하기.py
def dfs(x, y):
    global temp

    temp += 1

    for idx in range(4):
        nx = x + dx[idx]
        ny = y + dy[idx]

        if 0 <= nx < M and 0 <= ny < N and not paper[nx][ny]:
            paper[nx][ny] = 1
            dfs(nx, ny)


from sys import *
setrecursionlimit(10**5)

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

M, N, K = map(int, input().split())

paper = [[0] * N for _ in range(M)]

for _ in range(K):
    x1, y1, x2, y2 = map(int, input().split())

    for j in range(x1, x2):
        for i in range(y1, y2):
            paper[i][j] = 1

answer = 0
result = []
temp = 0

for i in range(M):
    for j in range(N):
        if paper[i][j] == 0:
            temp = 0
            answer += 1
            paper[i][j] = 1
            dfs(i, j)
            result.append(temp)

print(answer)
print(*sorted(result))

File: 백준/Silver I/2583. 영역 구하기/영역 구하기.cc
#include <iostream>
#include <algorithm>
using namespace std;

int M, N, K;
int paper[101][101];
int answer = 0;
int temp;
int result[10000];
int result_size = 0;

int dx[] = { 1, 0, -1, 0 };
int dy[] = { 0, 1, 0, -1 };

void dfs(int x, int y) {
	temp += 1;

	int nx, ny;

	for (int i = 0; i < 4; i++)
	{
		nx = x + dx[i];
		ny = y + dy[i];

		if (nx >= 0 && nx < M && ny >= 0 && ny < N && paper[nx][ny] == 0) {
			paper[nx][ny] = 1;
			dfs(nx, ny);
		}
	}
}

int main(void) {
	cin.tie(0);
	cout.tie(0);
	ios_base::sync_with_stdio(false);

	cin >> M >> N >> K;

	for (int i = 0; i < K; i++)
	{
		int x1, y1, x2, y2;
		cin >> x1 >> y1 >> x2 >> y2;

		for (int j = x1; j < x2; j++)
		{
			for (int k = y1; k < y2; k++)
			{
				paper[k][j] = 1;
			}
		}
	}

	for (int i = 0; i < M; i++)
	{
		for (int j = 0; j < N; j++)
		{
			if (paper[i][j] == 0) {
				answer++;
				paper[i][j] = 1;
				temp = 0;
				dfs(i, j);
				result[result_size++] = temp;
			}
		}
	}

	sort(result, result + result_size);

	cout << answer << "\n";
	
	for (int i = 0; i < result_size; i++)
	{
		cout << result[i] << " ";
	}

    return 0;
}